pydis_site/apps/api/viewsets/bot/mailing_list.py

```diff
@@ -0,0 +1,85 @@
+from django.db import IntegrityError
+from rest_framework import status
+from rest_framework.decorators import action
+from rest_framework.exceptions import ParseError
+from rest_framework.mixins import ListModelMixin, RetrieveModelMixin
+from rest_framework.request import Request
+from rest_framework.response import Response
+from rest_framework.viewsets import GenericViewSet
+
+from pydis_site.apps.api.serializers import MailingListSerializer
+from pydis_site.apps.api.models import MailingList, MailingListSeenItem
+
+
+class MailingListViewSet(GenericViewSet, ListModelMixin, RetrieveModelMixin):
+    """
+    View providing management and updates of mailing lists and their seen items.
+
+    ## Routes
+
+    ### GET /bot/mailing-lists
+    Returns all the mailing lists and their seen items.
+
+    #### Response format
+    >>> [
+    ...     {
+    ...         'id': 1,
+    ...         'name': 'python-dev',
+    ...         'seen_items': [
+    ...             {'hash': 'd81gg90290la8'},
+    ...             ...
+    ...         ]
+    ...     },
+    ...     ...
+    ... ]
+
+    ### GET /bot/mailing-lists/<name:str>
+    Retrieve a single mailing list and its seen items.
+
+    #### Response format
+    >>> {
+    ...     'id': 1,
+    ...     'name': 'python-dev',
+    ...     'seen_items': [
+    ...         {'hash': 'd81gg90290la8'},
+    ...         ...
+    ...     ]
+    ... }
+
+    ### POST /bot/mailing-lists/<name:str>/seen-items
+    Add a single seen item to the given mailing list. The request body should
+    be the hash of the seen item to add, as a plain string.
+
+    #### Request body
+    >>> str
+
+    #### Response format
+    Empty response.
+
+    #### Status codes
+    - 204: on successful creation of the seen item
+    - 400: if the request data was invalid
+    - 404: when the mailing list with the given name could not be found
+    """
+
+    lookup_field = 'name'
+    serializer_class = MailingListSerializer
+    queryset = MailingList.objects.prefetch_related('seen_items')
+
+    @action(detail=True, methods=["POST"],
+            name="Add a seen item for a mailing list", url_name='seen-items', url_path='seen-items')
+    def add_seen_item(self, request: Request, name: str) -> Response:
+        """Add a single seen item to the given mailing list."""
+        list_ = self.get_object()
```

Comment:
We can void the unnecessary query by doing the `isinstance` check before fetching the object

Reply:
Good idea, changed.